### ISSUE TITLE
False positive for RedundantCondition in closure

### ISSUE THREAD

**Issue by kkmuffme** [vimeo/psalm#2461]
ERROR: RedundantCondition - 16:12 - Found a redundant condition when evaluating $protocol and trying to reconcile type 'string(example)' to !empty

But the condition isn't redundant, as I need to be able to check if it's empty or not.

Minimal code to reproduce (on https://psalm.dev/)

```php
<?php
$protocol = 'foo';
$input = 'hello world';

$output = preg_replace_callback(
	"@(\w+)@",
	/**
	 * modifies the file paths to contain a protocol
	 *
	 * @param array $matches the matches we got
	 * @param string $protocol the file protocol prefix to use
	 *
	 * @return string without modification
	 */
	function ( $matches ) use ( $protocol ) {
		if ( !empty( $protocol ) ) {
			return $protocol . $matches[1];
		}

		// normal file protocol
		return $matches[1];
	},
	$input
);

echo $output;
```

**Comment by muglug** [comment]
Hey! This is actually a true positive, on account of `$protocol` is never empty in your example.

If protocol is variable then Psalm doesn't complain: https://psalm.dev/r/b5ae025628

**Comment by kkmuffme** [comment]
Nope, that's not true. Set $protocol = ''; and you will get that error too.

```php
<?php
$protocol = '';
$input = 'hello world';

$output = preg_replace_callback(
	"@(\w+)@",
	/**
	 * modifies the file paths to contain a protocol
	 *
	 * @param array $matches the matches we got
	 * @param string $protocol the file protocol prefix to use
	 *
	 * @return string without modification
	 */
	function ( $matches ) use ( $protocol ) {
		if ( !empty( $protocol ) ) {
			return $protocol . $matches[1];
		}

		// normal file protocol
		return $matches[1];
	},
	$input
);

echo $output;
```

**Comment by muglug** [comment]
No, you get a different error : https://psalm.dev/r/fa06217135 vs https://psalm.dev/r/efa0ff5605

`RedundantCondition` vs `TypeDoesNotContainType`

**Comment by kkmuffme** [comment]
Oh right, but how would I fix the TypeDoesNotContainType here? Since $protocol is an empty string, why does it give an error for this?

**Comment by muglug** [comment]
because you're checking that an empty string is not empty – that's clearly a bug

**Comment by kkmuffme** [comment]
Ok thx, get it now
